kse/src/org/kse/gui/dialogs/DGeneratingDHParameters.java

```diff
@@ -0,0 +1,192 @@
+import java.security.AlgorithmParameterGenerator;
+import java.security.AlgorithmParameters;
+import java.security.SecureRandom;
+import java.util.ResourceBundle;
+
+import javax.crypto.spec.DHParameterSpec;
+import javax.swing.AbstractAction;
+import javax.swing.ImageIcon;
+import javax.swing.JButton;
+import javax.swing.JComponent;
+import javax.swing.JFrame;
+import javax.swing.JLabel;
+import javax.swing.JProgressBar;
+import javax.swing.KeyStroke;
+import javax.swing.SwingUtilities;
+
+import org.bouncycastle.asn1.ASN1EncodableVector;
+import org.bouncycastle.asn1.ASN1Encoding;
+import org.bouncycastle.asn1.ASN1Integer;
+import org.bouncycastle.asn1.DERSequence;
+import org.bouncycastle.crypto.params.DHParameters;
+import org.bouncycastle.jce.provider.BouncyCastleProvider;
+import org.kse.gui.JEscDialog;
+import org.kse.gui.error.DError;
+
+import net.miginfocom.swing.MigLayout;
+
+/**
+ * Generates DH Parameters which the user may cancel at any time by pressing the
+ * cancel button.
+ *
+ */
+public class DGeneratingDHParameters extends JEscDialog {
+	private static final long serialVersionUID = 1L;
+
+	private static ResourceBundle res = ResourceBundle.getBundle("org/kse/gui/dialogs/resources");
+
+	private static final String CANCEL_KEY = "CANCEL_KEY";
+
+	private JLabel jlGenDHParameters;
+	private JProgressBar jpbGenDHParameters;
+	private JButton jbCancel;
+
+	private byte[] dhParameters;
+	private int keySize;
+	private Thread generator;
+	private boolean successStatus = true;
+
+	/**
+	 * Creates a new DGeneratingDHParameters dialog.
+	 *
+	 * @param parent
+	 *            The parent frame
+	 * @param keySize
+	 *            The key size to generate
+	 */
+	public DGeneratingDHParameters(JFrame parent, int keySize) {
+		super(parent, Dialog.ModalityType.DOCUMENT_MODAL);
+		this.keySize = keySize;
+		initComponents();
+	}
+
+	private void initComponents() {
+		//TODO Generate DH Parameters icon
+		jlGenDHParameters = new JLabel(res.getString("DGeneratingDHParameters.jlGenDHParameters.text"));
+		ImageIcon icon = new ImageIcon(getClass().getResource("images/genkp.png"));
+		jlGenDHParameters.setIcon(icon);
+
+		jpbGenDHParameters = new JProgressBar();
+		jpbGenDHParameters.setIndeterminate(true);
+		
+		jbCancel = new JButton(res.getString("DGeneratingDHParameters.jbCancel.text"));
+		jbCancel.addActionListener(evt -> cancelPressed());
+		jbCancel.getInputMap(JComponent.WHEN_IN_FOCUSED_WINDOW).put(KeyStroke.getKeyStroke(KeyEvent.VK_ESCAPE, 0),
+				CANCEL_KEY);
+		jbCancel.getActionMap().put(CANCEL_KEY, new AbstractAction() {
+			private static final long serialVersionUID = 1L;
+
+			@Override
+			public void actionPerformed(ActionEvent evt) {
+				cancelPressed();
+			}
+		});
+
+		Container pane = getContentPane();
+		pane.setLayout(new MigLayout("insets 10", "[]60","[]20[]20"));
+		pane.add(jlGenDHParameters, "wrap");
+		pane.add(jpbGenDHParameters, "growx, wrap");
+		pane.add(jbCancel, "align right");
```

Comment:
"[]60" means: put 60 pixels on the right side of every column and because there is only one column here, it creates that empty space on the right:

![grafik](https://user-images.githubusercontent.com/7840192/120906168-773ef080-c657-11eb-8f9a-b981865d0e00.png)

It's also most of the time better to use the predefined sizes like "unrelated" or "paragraph". They are converted by Miglayout into platform specific defaults.

```
		pane.setLayout(new MigLayout("insets dialog", "[]","[]unrel"));
		pane.add(jlGenDHParameters, "wrap");
		pane.add(jpbGenDHParameters, "growx, wrap");
		pane.add(jbCancel, "tag cancel");
```

Reply:
Updated, but please let me know if you think it needs further adjustment.

Reply:
Oh, you are right, there is still a little gap on the right side. That's because the dialog title is longer than the content. That can be fixed by saying that MigLayout shall fill the available space with the content: 
pane.setLayout(new MigLayout("insets dialog, fill",
As this is the last remaining open comment, I'll just merge the PR and do it afterwards myself.